pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -2800,12 +2806,34 @@ private CompletableFuture<Void> doTransactionAcknowledgeForResponse(MessageId me
             } else {
                 bitSetRecyclable.clear(messageIdAdv.getBatchIndex());
             }
-            cmd = Commands.newAck(consumerId, ledgerId, entryId, bitSetRecyclable, ackType, validationError, properties,
-                    txnID.getLeastSigBits(), txnID.getMostSigBits(), requestId, messageIdAdv.getBatchSize());
+            cmdList.add(Commands.newAck(consumerId, ledgerId, entryId, bitSetRecyclable, ackType, validationError, properties,
+                    txnID.getLeastSigBits(), txnID.getMostSigBits(), requestId, messageIdAdv.getBatchSize()));
```

Comment:
We can use one Ack command to ack multiple individual messages https://github.com/apache/pulsar/blob/99e6aebb88c8ea82c8b8d65261ce71f0ab8d36d8/pulsar-common/src/main/proto/PulsarApi.proto#L563

Why do we need to create multiple ack commands?

Reply:
There is only one ack command for batch messages. Adding it to cmdList is to unify the form.

Reply:
This is indeed an issue, but it is unrelated to this PR. This PR does not alter the behavioral logic of cumulative ack and batch messages. It simply consolidates the multiple commands of the chunk message into a list. This change does not cause any additional system calls to be triggered.

Reply:
Oh, sorry. I misunderstand this part.
My suggestion is to use the SingtonList and the ArrayList instead of using the LinkedList.
It will make the code easier to understand.

```diff
diff --git a/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java b/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java
index 0c17626484..e848a94402 100644
--- a/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java
+++ b/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java
@@ -2796,7 +2796,7 @@ public class ConsumerImpl<T> extends ConsumerBase<T> implements ConnectionHandle
         final MessageIdAdv messageIdAdv = (MessageIdAdv) messageId;
         final long ledgerId = messageIdAdv.getLedgerId();
         final long entryId = messageIdAdv.getEntryId();
-        final List<ByteBuf> cmdList = new LinkedList<>();
+        final List<ByteBuf> cmdList;
         if (MessageIdAdvUtils.isBatch(messageIdAdv)) {
             BitSetRecyclable bitSetRecyclable = BitSetRecyclable.create();
             bitSetRecyclable.set(0, messageIdAdv.getBatchSize());
@@ -2806,15 +2806,17 @@ public class ConsumerImpl<T> extends ConsumerBase<T> implements ConnectionHandle
             } else {
                 bitSetRecyclable.clear(messageIdAdv.getBatchIndex());
             }
-            cmdList.add(Commands.newAck(consumerId, ledgerId, entryId, bitSetRecyclable, ackType, validationError, properties,
-                    txnID.getLeastSigBits(), txnID.getMostSigBits(), requestId, messageIdAdv.getBatchSize()));
+            cmdList = Collections.singletonList(Commands.newAck(consumerId, ledgerId, entryId, bitSetRecyclable,
+                    ackType, validationError, properties, txnID.getLeastSigBits(), txnID.getMostSigBits(), requestId,
+                    messageIdAdv.getBatchSize()));
             bitSetRecyclable.recycle();
         } else {
             MessageIdImpl[] chunkMsgIds = this.unAckedChunkedMessageIdSequenceMap.remove(messageIdAdv);
             // cumulative ack chunk by the last messageId
             if (chunkMsgIds == null || ackType == AckType.Cumulative) {
-                cmdList.add(Commands.newAck(consumerId, ledgerId, entryId, null, ackType, validationError,
-                        properties, txnID.getLeastSigBits(), txnID.getMostSigBits(), requestId));
+                cmdList = Collections.singletonList(Commands.newAck(consumerId, ledgerId, entryId, null,
+                        ackType, validationError, properties, txnID.getLeastSigBits(), txnID.getMostSigBits(),
+                        requestId));
             } else {
                 if (Commands.peerSupportsMultiMessageAcknowledgment(
                         getClientCnx().getRemoteEndpointProtocolVersion())) {
@@ -2825,8 +2827,10 @@ public class ConsumerImpl<T> extends ConsumerBase<T> implements ConnectionHandle
                             entriesToAck.add(Triple.of(cMsgId.getLedgerId(), cMsgId.getEntryId(), null));
                         }
                     }
-                    cmdList.add(newMultiTransactionMessageAck(consumerId, txnID, entriesToAck, requestId));
+                    cmdList = Collections.singletonList(newMultiTransactionMessageAck(consumerId, txnID, entriesToAck,
+                            requestId));
                 } else {
+                    cmdList = new ArrayList<>();
                     for (MessageIdImpl cMsgId : chunkMsgIds) {
                         cmdList.add(Commands.newAck(consumerId, cMsgId.ledgerId, cMsgId.entryId, null, ackType,
                                 validationError, properties,
@@ -2854,7 +2858,7 @@ public class ConsumerImpl<T> extends ConsumerBase<T> implements ConnectionHandle
     }
 
     private ByteBuf newMultiTransactionMessageAck(long consumerId, TxnID txnID,
-                                                  List<Triple<Long, Long, ConcurrentBitSetRecyclable>> entries, long requestID) {
+                      List<Triple<Long, Long, ConcurrentBitSetRecyclable>> entries, long requestID) {
         BaseCommand cmd = newMultiMessageAckCommon(entries);
         cmd.getAck()
                 .setConsumerId(consumerId)

```